cpp/src/io/orc/writer_impl.cu

```diff
@@ -789,22 +849,26 @@ std::vector<StripeInformation> writer::impl::gather_stripes(
   return stripes;
 }
 
+void set_stat_desc_leaf_cols(device_span<orc_column_device_view const> columns,
+                             device_span<stats_column_desc> stat_desc,
+                             rmm::cuda_stream_view stream)
+{
+  thrust::for_each(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator(0ul),
+    thrust::make_counting_iterator(stat_desc.size()),
+    [=] __device__(auto idx) { stat_desc[idx].leaf_column = &columns[idx].cudf_column; });
+}
+
 std::vector<std::vector<uint8_t>> writer::impl::gather_statistic_blobs(
-  const table_device_view& table,
-  host_span<orc_column_view const> columns,
-  host_span<stripe_rowgroups const> stripe_bounds)
+  orc_table_view const& orc_table, file_segmentation const& segmentation)
 {
-  auto const num_rowgroups = stripes_size(stripe_bounds);
-  size_t num_stat_blobs    = (1 + stripe_bounds.size()) * columns.size();
-  size_t num_chunks        = num_rowgroups * columns.size();
+  auto const num_stat_blobs = (1 + segmentation.num_stripes()) * orc_table.num_columns();
 
-  std::vector<std::vector<uint8_t>> stat_blobs(num_stat_blobs);
-  hostdevice_vector<stats_column_desc> stat_desc(columns.size(), stream);
+  hostdevice_vector<stats_column_desc> stat_desc(orc_table.num_columns(), stream);
   hostdevice_vector<statistics_merge_group> stat_merge(num_stat_blobs, stream);
```

Comment:
This is actually pretty messy to convert, the array is used in creative ways. 
If you don't mind, I'd prefer to do this is a separate PR (since there's a few more improvements that can be made in this code).